[Fix] Removed unnecessarily changing tuple output to list output

zli11010 opened the pull request:
## 📥 Pull Request

**📘 Description**
Fixed this bug: https://github.com/AgentOps-AI/agentops/issues/607

**🧪 Testing**
Ran my code that uses AgentOps and now it works. (Changing list to tuple doesn't affect what gets logged.)

zli11010 committed:
[Fix] Removed unnecessarily changing tuple output to list output

areibman commented on agentops/decorators.py:
```diff
@@ -67,10 +67,6 @@ async def async_wrapper(*args, session: Optional[Session] = None, **kwargs):
                 try:
                     returns = await func(*args, **kwargs)
 
-                    # If the function returns multiple values, record them all in the same event
-                    if isinstance(returns, tuple):
-                        returns = list(returns)
-
```
Instead of removing these functions, we can instead do what we intended at fist-- just return a tuple

```
if istinatnce(returns, tuple):
    returns = tuple(returns)
```

Removing these lines of code altogether will have unintended consequences.

zli11010 replied:
May I ask what the suggested change does? It seems (in most cases) equivalent to removing the two lines of code.

Unfortunately your suggested change would still break the code I'm using, since the code checks the output of the function is an instance of a *specific subclass* of `tuple` (more specifically, a custom `namedtuple`). Maybe one could argue that this is bad practice, but I believe it's more important that a decorator that does logging doesn't change the output of the function.

teocns replied:
> but AFAIK the only reason the tuple was being cast to a list was so that it can be dumped into a JSON, but json.dumps actually handles that automatically.

Indeed that should be handled by a `JSONEncoder` which is where these sorts of casts belong

areibman commented:
Left a review @zli11010-- should be an easy fix! Please let me know if it still causes problems with the repo you're trying

teocns commented:
@areibman can we have more documentation around the expected functionality of `returns`? I too had a hard time figuring out, maybe some docstrings will do 👍

zli11010 committed:
Revert "[Fix] Removed unnecessarily changing tuple output to list output"

This reverts commit 829fa433db7e0f754233220ca45b4943b7ad1a7d.

zli11010 committed:
Fix multiple returns being modified by decorator

zli11010 commented:
@areibman Since you had concerns about deleting the lines, I tried to keep the original semantics with the only change being fixing the fact that the type casting modified the output of the original function. Hopefully that's better!

dot-agi committed:
Merge branch 'main' into fix/tuple_changed_to_list

dot-agi committed:
Merge branch 'main' into fix/tuple_changed_to_list

dot-agi commented:
What's the status of this PR?

areibman commented:
@teocns can you take a stab at completing this? Just want to make sure we don’t have any downstream effects if we merge

teocns committed:
Merge branch 'main' into fix/tuple_changed_to_list

teocns committed:
test: Add tests to ensure @record_* decorators don't change return types

Added tests to verify PR #608, which fixes decorators changing tuple outputs
into lists. Checks:

1. Decorators keep the original return type for:
   - Simple tuples like (1, "test")
   - Named tuples like Point(x=1, y=2)
   - Nested tuples containing multiple values
   - Custom tuple types (fixing bug #607)

2. Tests also show that JSON handles tuples correctly without needing
   to convert them to lists first, proving the point that the
   conversion was unnecessary.

teocns commented:
The removal of the tuple-to-list conversion seems to be the right approach; the conversion was actually unnecessary since:

- Our serialization layer (`safe_serialize` and `filter_unjsonable`) already handles tuples correctly
- The conversion was happening too early (at decorator level) rather than at serialization time

Added some tests to ensure `@record_*` decorators behave correctly in regards to the issue, especially proving the json serialization utilities already handle tuple types correctly without needing to convert them to lists first, proving the point that the conversion was unnecessary :) 
   
   
If this gets merged, we should open a ticket to unify:
- [`test_record_action.py`](https://github.com/AgentOps-AI/agentops/blob/80d4bf59cc7b0faef43b8e251b2002a3ee40f65f/tests/test_record_action.py)
- [`test_record_tool.py`](https://github.com/AgentOps-AI/agentops/blob/80d4bf59cc7b0faef43b8e251b2002a3ee40f65f/tests/test_record_tool.py)

into:
- [`test_decorators.py`](https://github.com/AgentOps-AI/agentops/blob/64fc5ab7c4a8e3e97394a195e1394b5c381a17a3/tests/test_decorators.py)

--- 

> Rationale: Both `record_action` and `record_tool` are decorators with similar functionality and testing patterns.
Having their tests in separate files creates unnecessary fragmentation.

teocns committed:
Merge branch 'main' into fix/tuple_changed_to_list

dot-agi committed:
Merge branch 'main' into fix/tuple_changed_to_list

teocns committed:
Merge branch 'main' into fix/tuple_changed_to_list
